Aggregate functions on Trends (closes #1193)

Twixes opened the pull request:
## Changes

This adds Sum, Average, Minimum and Maximum to Total (now called Total volume) and DAU. These new options run aggregate functions on event properties, so that it's possible to chart e.g. average purchase total on each day.

![Math selection](https://user-images.githubusercontent.com/4550621/87836400-33e96100-c890-11ea-9924-ae48e1c0e417.png)
![Math property selection](https://user-images.githubusercontent.com/4550621/87836402-35b32480-c890-11ea-9203-1822b01997a4.png)

## Checklist

- [x] All querysets/queries filter by Team (if applicable)
- [x] Backend tests (if applicable)
- [x] Cypress E2E tests (if applicable)

Twixes committed:
Update entityFilterLogic for filter.math_property

Twixes committed:
Add new math types and math_property to ActionFilterRow

Twixes committed:
Add backend support for new math types

Twixes committed:
Update data-attrs

Twixes committed:
Add dropdown Cy test

Twixes committed:
Add API tests

Twixes committed:
Improve process_math style

Twixes committed:
Add resiliency to non-number property value occurences

Twixes committed:
Merge branch '1193-trends-math' of https://github.com/PostHog/posthog into 1193-trends-math

Twixes committed:
Annotate for clarity

Twixes committed:
Update UI text

Twixes committed:
Update UI text a bit more

Twixes committed:
Reword tooltip

Twixes committed:
Update UI slightly

Twixes committed:
Update capitalization

Twixes committed:
Improve internal style

EDsCODE commented on frontend/src/scenes/trends/ActionFilter/ActionFilterRow.js:
```diff
@@ -111,36 +207,59 @@ export function ActionFilterRow({ logic, filter, index, hideMathSelector }) {
+            data-attr={`math-property-selector-${props.index}`}
+        >
+            {props.properties
+                .filter(({ value }) => value[0] !== '$' && value !== 'distinct_id')
```
we should filter for the property type here. i'm guessing the common use case in custom properties will actually be numerical but it will be more comprehensive and intuitive if we remove properties that might be booleans for example so it won't look like you can average a boolean

Twixes replied:
I considered this at first, but then there is no typing in properties, so we need to gracefully handle a situation like: user releases their software with an oversight where a string is accidentally sent in place of a number, user pushes a hotfix, but 100 events with that string remain – in this case we still should be able to count all the other, proper events. So properties that *might* be booleans should be shown. After a bit of rethinking though, it'd be good to hide properties that *never* are numbers (to date). I'll put that in.

EDsCODE commented on frontend/src/layout/LatestVersion.js:
```diff
@@ -4,7 +4,7 @@ import api from './../lib/api'
 import { ChangelogModal } from '~/layout/ChangelogModal'
 import { userLogic } from 'scenes/userLogic'
-import { CheckOutlined, WarningOutlined } from '@ant-design/icons'
+import { CheckOutlined, BulbOutlined } from '@ant-design/icons'
```
These changes were already merged in #1216

Twixes replied:
Yup, they were, I rebased to work on master at the time and this got in. Shouldn't be a conflict though.

EDsCODE reviewed (commented):
Didn't do a full review. Just did a quick look through and it looks good! There's a small change I'd suggest. Will do a full review and QA tomorrow pending your response

Twixes commented:
Only event properties which have at least 1 `number` type occurrence are aggregate function options now.
This has a downside in that the list of them needs to be computed beforehand – might warrant testing the performance hit on a large database.

Twixes commented:
Well, this going through all events to check property types is surely a bad idea on a large database. The alternative is adding a property type check of this sort at event insertion time, although that would unfortunately exclude properties on events from before this feature from being used in aggregate functions.

EDsCODE commented:
I would suggest the property type check on insertion time but like you noted it will be lacking on previous properties and we'll have to figure out a solution to retroactively add the types then. I think since that's out scope for this PR, no property type check will be sufficient for this PR

Twixes committed:
Improve code style

Twixes committed:
Filter out property `token`

Twixes committed:
Utilize formatted string

EDsCODE commented:
One last UX bug I found is that when there are no properties you still get an empty dropdown. Might be better to have the field say "no properties" or some sort of notice

<img width="513" alt="Screen Shot 2020-07-21 at 9 02 01 AM" src="https://user-images.githubusercontent.com/13127476/88057978-dd4e8200-cb30-11ea-8f2c-f33cc948353c.png">

Twixes committed:
Don't show empty dropdown contents

Twixes committed:
Add event_properties_numerical for the future

Twixes committed:
Fix dropdown button title

Twixes commented:
Good point, added empty children handling to Dropdown.
Regarding only showing numerical properties, I added `event_properties_numerical`, updated at insertion time same way as `event_properties`. I think this should allow us to restrict options here somewhere down the road, when this check will have already been processing our users' events for a while. Does that make sense?

Twixes committed:
Merge branch 'master' into 1193-trends-math
